### ISSUE TITLE
System.Enum doesn't map to class when using ProjectTo

### ISSUE THREAD

**Issue by justintconroy** [AutoMapper/AutoMapper#2552]
I'm trying to convert enum values (for any enum in my project) to a special view model called `NamedEntityDto`. It's easy enough to create the mapping, and it works when you just call `Map` on an object containing that object, but it doesn't work if you call `ProjectTo` with that same object. My tests show that the AutoMapper considers the configuration to be valid.

### Source/destination types

```csharp
System.Enum

public enum MyEnum
{
    one = 1,
    two = 2,
    three = 3
}

public enum OtherEnum
{
    A = 1,
    B = 2,
    C = 3
}

public class NamedEntityDto
{
    public int Id { get; set; }
    public string Name { get; set; }
}

public class MyEntity
{
    public int Id { get; set; }
    public MyEnum MyEnum { get; set; }
    public OtherEnum OtherEnum { get; set; }
}

public class MyEntityDto
{
    public int Id { get; set; }
    public NamedEntityDto MyEnum { get; set; }
    public NamedEntityDto OtherEnum { get; set; }
}

```

### Mapping configuration

```csharp
CreateMap<System.Enum, NamedEntityDto>()
    .ForMember(dest => dest.Id,
                          opt => opt.MapFrom(src => src))
    .ForMember(dest => dest.Name,
                           opt => opt.MapFrom(src => src.ToString()));

CreateMap<MyEntity, MyEntityDto>();
```

### Version: x.y.z
 
* AutoMapper **6.2.2**
* EntityFramework **6.2.0**

### Expected behavior

When calling `ProjectTo` for `MyEntity`, the enum values should be mapped to `NamedEntityDto`s in the `MyEntityDto`.

### Actual behavior

AutoMapper throws this exception

```
AutoMapper.AutoMapperMappingException:'Unable to create a map expression from . (MyEnum) to NamedEntityDto.Id (System.Int32)'
```

### Steps to reproduce

```csharp
public class MyEntityController
{
    public IActionResult Get()
    {
        var myEntities = _dbContext.MyEntities;

        // This works
       var result =  _mapper.Map<MyEntityDto>(myEntities.FirstOrDefault());

        // This doesn't work.
        var list = myEntities.ProjectTo<MyEntityDto>().ToList();

        return Ok();
    }
}
```

I also tried explicitly converting the `Enum` to `Int32` in the mapping:

```csharp
CreateMap<System.Enum, NamedEntityDto>()
    .ForMember(dest => dest.Id,
                          opt => opt.MapFrom(src => Convert.ToInt32(src)))
    .ForMember(dest => dest.Name,
                           opt => opt.MapFrom(src => src.ToString()));
```

The result is more or less the same, but the exception is different:

```
System.Argument Exception: 'Expression of type 'MyEnum' cannot be used for parameter of type 'System.Object' of method 'Int32 ToInt32(System.Object)'
```

You can't directly cast it either (i.e. `(int)src`) because it's working on the type `System.Enum` rather than the actual enums.

**Comment by lbargaoanu** [comment]
Try to write a LINQ query that does what you want.

**Comment by justintconroy** [comment]
Hmm, it doesn't seem like I can do that... I guess that means I can't use `ProjectTo` here. I tried this:

```csharp
var linqResult = myEntities.Select(e => new MyEntityDto
    {
        Id = e.Id,
        MyEnum = new NamedEntityDto()
            {
                Id = e.MyEnum,
                Name = e.MyEnum.ToString(),
            },
                OtherEnum = new NamedEntityDto()
                    {
                        Id = e.OtherEnum,
                        Name = e.OtherEnum.ToString(),
                    }
            }).ToList();
```
But the compiler gives an error that looks very similar to the exception thrown:

```
Cannot implicitly convert type 'MyEnum' to 'int'. An explicit conversion exists (are you missing a cast?)
```

Explicitly casting the value to `int` in the LINQ query works, but of course it doesn't work when operating on just `System.Enum`. So it looks like maybe I'm stuck. 😟 

I guess I can find a workaround that converts the values after the query is finished, I was just hopeful that I could take care of it once with AutoMapper and have it work for any enum. Thank you for pointing me in the right direction anyways.

**Comment by lbargaoanu** [comment]
I believe that would work if we made this conversion built in. So if you're into expressions, you're welcome to try a PR. We'll help. But you also have ToString in there. Last time I checked EF6 doesn't support that.

**Comment by justintconroy** [comment]
I guess I hadn't gotten far enough to see the `ToString` fail. :-P

What would the built-in conversion need to be? `Enum` to `int`? I suppose if the `ToString` isn't going to work, it's not all that helpful in my case anyways.

**Comment by lbargaoanu** [comment]
Yes, Enum to the underlying type, like Map. About ToString, do test, I could be wrong :)

**Comment by justintconroy** [comment]
Well, the following works:

```csharp
var linq = myEntities
    .Select(e => new MyEntityDto
    {
        Id = e.Id,
        MyEnum = new NamedEntityDto()
        {
            Id = (int) e.MyEnum,
            Name = e.MyEnum.ToString(),
        },
        OtherEnum = new NamedEntityDto()
        {
            Id = (int) e.OtherEnum,
            Name = e.OtherEnum.ToString(),
        }
    }).ToList();
```

I don't think that's quite the same thing though as calling `ToString` on `System.Enum`. This mapping, for example, does not work:

```csharp
CreateMap<Enum, NamedEntityDto>()
    .ForMember(dest => dest.Id,
                opt => opt.Ignore())
                //opt => opt.MapFrom(src => src))
    .ForMember(dest => dest.Name,
                opt => opt.MapFrom(src => src.ToString()));
```

Which gives the exception:

```
System.InvalidOperationException: 'The binary operator NotEqual is not defined for the types 'MyEnum' and 'System.Object'.'
```

Unfortunately, I don't think I can afford to spend much more time on this at work. I might be able to come back to it this weekend. What should I attempt to poke as far as built-in conversions go?

**Comment by lbargaoanu** [comment]
You would need to add another one [here](https://github.com/AutoMapper/AutoMapper/blob/b3b4ed6223d50c9c083e898e0f30457292f3fc7f/src/AutoMapper/QueryableExtensions/ExpressionBuilder.cs#L42). And a test in AM.IntegrationTests.

**Comment by lbargaoanu** [comment]
ToString doesn't work for me with EF 6.2.

**Comment by justintconroy** [comment]
I've been testing with EF 6.2.

**Comment by lbargaoanu** [comment]
It works with a LINQ query, it doesn't work with ProjectTo although the expression looks good. Weird.

**Comment by justintconroy** [comment]
## Potential Workarounds
I tried a couple different things. I'll post them here for the benefit of anyone who might find this issue from a search engine.

First, I tried to change up the `NamedEntityDto` by having a real `EnumValue` for the enum to map to, and `get` methods for the properties that I actually want to see:

```csharp
public class NamedEnumDto
{
    public Enum EnumValue { get; set; }

    public int Id
    {
        get
        {
            return Convert.ToInt32(EnumValue);
        }
    }

    public string Name
    {
        get
        {
            return EnumValue.ToString();
        }
    }
}

CreateMap<Enum, NamedEnumDto>()
    .ForMember(dest => dest.EnumValue,
                opt => opt.MapFrom(src => src))
    .ForAllOtherMembers(opt => opt.Ignore());
```

This would be an acceptable workaround for me because the ultimate goal is to serialize these object to JSON. I can just instruct the JSON formatter to ignore the `EnumValue` property. I can also just stick `NamedEnumValue` directly into my DTO with no special handling.

Unfortunately this doesn't work. When trying to project this mapping, we get an exception:

```
System.InvalidOperationException: 'The binary operator NotEqual is not defined for the types 'TestEnumModel.MyEnum' and 'System.Object'.'
```
The error message doesn't make a lot of sense to me, but I tried just doing a LINQ query directly:

```csharp
var linq = myEntities
    .Select(e => new MyEntityDto
    {
        Id = e.Id,
        MyEnum = new NamedEnumDto()
        {
            EnumValue = e.MyEnum
        },
        OtherEnum = new NamedEnumDto()
        {
            EnumValue = e.OtherEnum
        }
    }).ToList();
```

This also fails, and gives a slightly more meaningful exception:

```
System.InvalidOperationException: 'The specified cast from a materialized 'System.Int32' type to the 'System.Enum' type is not valid.'
```

So it looks like mapping enums to `System.Enum` is out of the running.

## Something that actually works.
It's hackish, and you'll have to do it for every enum in a DTO you want to use it on, so it's not simple to just drop in, but it works:

```csharp
public class MyEntityDto
{
    public int Id { get; set; }

    public MyEnum MyEnumField { get; set; }
    public NamedEntityDto MyEnum
    {
        get
        {
            return Mapper.Map<NamedEntityDto>(MyEnumField);
        }
    }

    public OtherEnum OtherEnumField { get; set; }
    public NamedEntityDto OtherEnum
    {
        get
        {
            return Mapper.Map<NamedEntityDto>(OtherEnumField);
        }
    }
}

CreateMap<MyEntity, MyEntityDto>();
    .ForMember(dest => dest.MyEnumField,
                opt => opt.MapFrom(src => src.MyEnum))
    .ForMember(dest => dest.MyEnum,
                opt => opt.Ignore())
    .ForMember(dest => dest.OtherEnumField,
                opt => opt.MapFrom(src => src.OtherEnum))
    .ForMember(dest => dest.OtherEnum,
                opt => opt.Ignore());

CreateMap<Enum, NamedEntityDto>()
    .ForMember(dest => dest.Id,
                opt => opt.MapFrom(src => src))
    .ForMember(dest => dest.Name,
                opt => opt.MapFrom(src => src.ToString()));
```

Basically just mapping to the same enum type as usual, and then exposing some properties with `get` methods to have the right types to serialize. It's hacky, and I don't like it, but it works. I'm going to use it for now, and hopefully try poking some of the things @lbargaoanu pointed me towards this weekend to see if I can get a better solution.

**Comment by lbargaoanu** [comment]
It seems to be Enum.ToString vs Object.ToString. I think in EF Core works as it is.
